core/trino-main/src/main/java/io/trino/operator/output/BytePositionsAppender.java

```diff
@@ -14,31 +14,239 @@
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.BlockUtil;
+import io.trino.spi.block.ByteArrayBlock;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.spi.block.BlockUtil.calculateBlockResetSize;
+import static java.lang.Math.max;
 
 public class BytePositionsAppender
-        implements PositionsAppender
+        implements BlockTypeAwarePositionsAppender
 {
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BytePositionsAppender.class).instanceSize();
+    private static final Block NULL_VALUE_BLOCK = new ByteArrayBlock(1, Optional.of(new boolean[] {true}), new byte[1]);
+
+    @Nullable
+    private final BlockBuilderStatus blockBuilderStatus;
+    private boolean initialized;
+    private final int initialEntryCount;
+
+    private int positionCount;
+    private boolean hasNullValue;
+    private boolean hasNonNullValue;
+
+    // it is assumed that these arrays are the same length
+    private boolean[] valueIsNull = new boolean[0];
+    private byte[] values = new byte[0];
+
+    private long retainedSizeInBytes;
+
+    public BytePositionsAppender(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this.blockBuilderStatus = blockBuilderStatus;
+        this.initialEntryCount = max(expectedEntries, 1);
+
+        updateDataSize();
+    }
+
+    @Override
+    public void appendTo(IntArrayList positions, Block block)
```

Comment:
Why `appendTo`? We are not appending to block here. Just `append` will suffice IMO